state-sync/state-sync-v1/src/client.rs

```diff
@@ -10,7 +10,8 @@ use futures::{
 
 /// Messages used by the StateSyncClient for communication with the StateSyncCoordinator.
 pub enum CoordinatorMessage {
-    GetSyncState(oneshot::Sender<SyncState>), // Return the local sync state.
+    DisableReadOnlyMode(oneshot::Sender<Result<(), Error>>), // Disables read only mode for state sync.
```

Comment:
hmm, dynamically disabling read-only mode seems pretty dangerous to me. it seems much safer to just restart the node with state-sync-v2 turned off, when you get the no-progress alerts.

Reply:
Here's my thinking:
1. Unless we want to break backward compatibility, we need to be able to run both simultaneously. Hence, the read-only mode for state sync v1.
2. The burden of asking the operator to restart the node when no progress is being made seems a little impractical. We know for certain that there will be a period where state sync v2 won't make any progress for various nodes, so expecting operators to manually restart and set a new config value doesn't sound great.
3. Even if we did require a manual restart, all of the logic here is based around a single flag, so if that flag is switched on dynamically (vs. after a restart), nothing is really going to change from the perspective of the code flows.
4. Looking at the risk: the tests cover the functionality pretty well. Moreover, we already had most of this *risk* due to the potential race conditions that exist with consensus.

So, all in, all, I'm not that worried about this change. 😄

Reply:
> 1. ... we need to be able to run both simultaneously. Hence, the read-only mode for state sync v1.

right, I think we're all on board here with running SSv1 in read-only mode for at least one release.

> 2. ... We know for certain that there will be a period where state sync v2 won't make any progress for various nodes, so expecting operators to manually restart and set a new config value doesn't sound great.

if we can stabilize the storage-service API at least one release before state-sync-v2, then we can roll out the storage-service ahead of time, so everyone is at least serving the content we need to support SSv2. then we won't have any downtime or inability to sync unless something is very wrong with SSv2.

even better, after rolling out the storage service in a release, we can test out SSv2 live with some full nodes, move up to the Association nodes or Novi validators, then push it out everywhere. that way we can build up our confidence that SSv2 is working as intended.

> 3. Even if we did require a manual restart, all of the logic here is based around a single flag, so if that flag is switched on dynamically (vs. after a restart), nothing is really going to change from the perspective of the code flows.

I guess my fear is the potential race conditions during startup and shutdown of each protocol (SSv1, SSv2, and Consensus). I'm not super familiar with the flows here so it might be a total non-issue, but my gut feeling is that runtime coordination b/w SSv1 and SSv2 introduces lots of fun and exciting and poorly-explored states into the state space : p

IMO it's much easier to reason about nodes that are _purely_ SSv1 driven or SSv2 driven but not some dynamic combination of both. in fact, I would say the same thing the existing SSv1 vs Consensus sync modes : )

Reply:
> if we can stabilize the storage-service API at least one release before state-sync-v2, then we can roll out the storage-service ahead of time, so everyone is at least serving the content we need to support SSv2.

Aah, yes, my implicit assumption was that we wouldn't have enough time to do this and we'd more than likely land everything at once. But, let me confirm this and get back to you. If we can layer the releases, then you're exactly right and this would be much nicer 😄 

> I guess my fear is the potential race conditions during startup and shutdown of each protocol (SSv1, SSv2, and Consensus).

Completely, agree. There is some added risk here, but I don't think it's as great as we might think because we're already facing this risk with consensus and SSv1. All this change would do (if the logic was buggy) is create a race where 3 writers might write to storage, instead of 2 that exist today. Note though, the chances of these races is very slim given what needs to happen to get there. It's so slim that a race existed between consensus and SSv1 for a long time and no one noticed 😄